cou/steps/plan.py

```diff
@@ -596,7 +595,27 @@ async def _get_upgradable_hypervisors_machines(
     if cli_force:
         return nova_compute_machines
 
-    return await get_empty_hypervisors(nova_compute_units, analysis_result.model)
+    empty_hypervisors = await get_empty_hypervisors(nova_compute_units, analysis_result.model)
+
+    machine_to_unit_name = {
+        machine.machine_id: unit.name
+        for unit, machine in zip(nova_compute_units, nova_compute_machines)
+    }
```

Comment:
I think you can change  `_get_nova_compute_units_and_machines` to be a dictionary contains machines object as keys and nova-compute units as values. It would be more handy than constructing the dictionary here

Reply:
I am confused... `_get_nova_compute_units_and_machines` returns
```
return nova_compute_units, [unit.machine for unit in nova_compute_units]
```
which seems kinda useless...
And then we zip them back together here (?)
...only to reconstruct the unit name from the machine id.

Isn't it overly convoluted?

`get_empty_hypervisors` is used only once (aside from here) so it's easy to refactor. If we change it to return the set of empty hypervisor _units_, all of this should simplify, I think, to

```
hypervisor_units, _ = _get_nova_compute_units_and_machines(
    analysis_result.apps_data_plane
) # or hypervisor_units = _get_nova_compute_units(analysis_result.apps_data_plane)
empty_hypervisor_units = await get_empty_hypervisors(hypervisor_units, analysis_result.model)
non_empty_hypervisor_units = hypervisor_units - empty_hypervisor_units # make them sets at some point

stringify_units = lambda x: ", ".join([u.name for u in x])

skipped_unit_names = stringify_units(non_empty_hypervisor_units)

# etc
```

Reply:
Should logos listed below be part of `get_empty_hypervisors` and like that, we would not need to have `machine_to_unit_name` or any issue what `get_empty_hypervisors` should return?

If we move it there, we should not use words like `Skipped` and `Selected`, but we can use `Found not-empty ...`. Also, the message with empty hypervisors should not be warning level?

Reply:
yes, logging within `get_empty_hypervisors` would be easier, but not fully correct from my point of view. The point here is informing the user that some hypervisors that they expected to upgrade will not actually be upgraded, therefore the logging should occur within the plan stage, not within a utils function.

Why do you think the log level should be 'warning'?